File: 2023/08.24/송기훈/친구네트워크.java
package baekjoon.gold.two;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Arrays;
import java.util.HashMap;
import java.util.StringTokenizer;

public class 친구네트워크 {
    // 유니온 파인드
    // https://ssungkang.tistory.com/entry/Algorithm-%EC%9C%A0%EB%8B%88%EC%98%A8-%ED%8C%8C%EC%9D%B8%EB%93%9CUnion-Find

    static int T, F;
    static int[] parent, count;

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        T = Integer.parseInt(br.readLine());
        for (int t = 0; t < T; t++) {
            F = Integer.parseInt(br.readLine());
            HashMap<String, Integer> names = new HashMap<>();
            parent = new int[F * 2];
            count = new int[F * 2];
            Arrays.setAll(parent, i -> i);
            Arrays.fill(count, 1);
            int index = 0;

            for (int f = 0; f < F; f++) {
                StringTokenizer st = new StringTokenizer(br.readLine());
                String a = st.nextToken();
                String b = st.nextToken();
                if (!names.containsKey(a)) {
                    names.put(a, index);
                    index += 1;
                }
                if (!names.containsKey(b)) {
                    names.put(b, index);
                    index += 1;
                }
                union(names.get(a), names.get(b));
                System.out.println(count[find(names.get(b))]);
            }
        }
    }

    static int find(int index) {
        // 타고 타고 올라가서 부모 노드를 찾는다
        if (parent[index] == index) {
            return index;
        } else {
            return parent[index] = find(parent[index]);
        }
    }

    static void union(int idxA, int idxB) {
        int parentA = find(idxA);
        int parentB = find(idxB);

        // 부모가 같으므로 이미 연결됨
        if (parentA == parentB) {
            return;
        }

        parent[Math.max(parentA, parentB)] = Math.min(parentA, parentB);
        count[Math.min(parentA, parentB)] += count[Math.max(parentA, parentB)];
    }

}

File: 2023/08.24/조수연/친구_네트워크.kt
import java.io.*
import java.util.*
import kotlin.collections.HashMap

lateinit var st: StringTokenizer
lateinit var relation: HashMap<String,String>
lateinit var count: HashMap<String,Int>

fun main() = with(BufferedReader(InputStreamReader(System.`in`))) {
    val bw = BufferedWriter(OutputStreamWriter(System.out))

    val t = readLine().toInt()

    repeat(t) {
        val f = readLine().toInt()
        relation = HashMap()
        count = HashMap()
        
        repeat(f) {
            st = StringTokenizer(readLine())

            val p1 = st.nextToken()
            val p2 = st.nextToken()

            relation.putIfAbsent(p1,p1)
            count.putIfAbsent(p1,1)

            relation.putIfAbsent(p2,p2)
            count.putIfAbsent(p2,1)

            union(p1,p2)

            bw.write("${count[find(p1)]!!}\n")
        }
    }

    bw.flush()
    bw.close()
}

/* 분리집합 */
fun union(x: String, y: String) {
    val nx = find(x)
    val ny = find(y)

    if (nx != ny) {
        relation[ny] = nx
        count.replace(nx,count[ny]!! + count[nx]!!)
    }
}

fun find(x: String): String {
    return if (relation[x] == x) x
    else {
        relation[x] = find(relation[x]!!)
        relation[x]!!
    }
}